packages/valory/skills/decision_maker_abci/policy.py

```diff
@@ -43,102 +43,138 @@ def argmax(li: List) -> int:
-        ]
+    def random_tool(self) -> str:
+        """Get the name of a tool randomly."""
+        return random.choice(list(self.accuracy_store.keys()))  # nosec
 
     @property
-    def best_tool(self) -> int:
+    def best_tool(self) -> str:
         """Get the best tool."""
-        return argmax(self.reward_rates)
-
-    def add_new_tools(self, indexes: List[int], avoid_shift: bool = False) -> None:
-        """Add new tools to the current policy."""
-        if avoid_shift:
-            indexes = sorted(indexes, reverse=True)
-
-        for i in indexes:
-            self.counts.insert(i, self.initial_value)
-            self.rewards.insert(i, float(self.initial_value))
-
-    def remove_tools(self, indexes: List[int], avoid_shift: bool = False) -> None:
-        """Remove the knowledge for the tools corresponding to the given indexes."""
-        if avoid_shift:
-            indexes = sorted(indexes, reverse=True)
-
-        for i in indexes:
-            try:
-                del self.counts[i]
-                del self.rewards[i]
-            except IndexError as exc:
-                error = "Attempted to remove tools using incorrect indexes!"
-                raise ValueError(error) from exc
-
-    def select_tool(self, randomness: RandomnessType) -> Optional[int]:
+        weighted_accuracy = list(self.weighted_accuracy.values())
+        best = argmax(weighted_accuracy)
+        return self.tools[best]
+
+    def update_weighted_accuracy(self) -> None:
+        """Update the weighted accuracy for each tool."""
+        self.weighted_accuracy = {
+            tool: acc_info.accuracy
+            * (acc_info.requests - acc_info.pending)
+            / self.n_requests
+            for tool, acc_info in self.accuracy_store.items()
+        }
+
+    def select_tool(self, randomness: RandomnessType) -> Optional[str]:
         """Select a Mech tool and return its index."""
         if self.n_tools == 0:
             return None
 
         random.seed(randomness)
-        if sum(self.reward_rates) == 0 or random.random() < self.eps:  # nosec
+        if not self.has_updated or random.random() < self.eps:  # nosec
             return self.random_tool
 
         return self.best_tool
 
-    def tool_used(self, index: int) -> None:
-        """Increase the times used for the tool corresponding to the given index."""
-        self.counts[index] += 1
+    def tool_used(self, tool: str) -> None:
+        """Increase the times used for the given tool."""
+        self.accuracy_store[tool].pending += 1
+        self.update_weighted_accuracy()
 
-    def add_reward(self, index: int, reward: float = 0) -> None:
-        """Add a reward for the tool corresponding to the given index."""
-        self.rewards[index] += reward
+    def reward(self, tool: str) -> None:
+        """Reward the given tool."""
+        self.accuracy_store[tool].requests += 1
+        self.accuracy_store[tool].pending -= 1
```

Comment:
having two queues for pending and closed requests makes the management more complicated. Why we cannot just have the number of requests?

Reply:
Because the calculation of the weighted accuracy uses both.